### PR TITLE
Observer:  add some metrics for the waiting queue of snapshot generator. 

### LINKED ISSUE #12945: Observer： snapshot should add more metrics for the waiting duration of snapshot generator.
## Development Task
When the duration of region operator is bigger, the waiting duration maybe biggger also. The metrics will be helpful to explain the slow reason.

image

### PR DESCRIPTION
**Pull request by bufferflies**
### What is changed and how it works?

Issue Number: Close #12945

What's Changed:

```commit-message
```

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
snap generate pending task 
image

snap limit consumer 

image

Side effects

### Release note 

```release-note
None.
```

### CODE REVIEW

File: components/raftstore/src/store/snap/io.rs [hunk 1]
```diff
@@ -180,12 +184,13 @@ where
         box_try!(sst_writer.into_inner().finish());
         box_try!(File::open(path).and_then(|f| f.sync_all()));
         info!(
-            "build_sst_cf_file_list builds {} files in cf {}. Total keys {}, total size {}. raw_size_per_file {}",
+            "build_sst_cf_file_list builds {} files in cf {}. Total keys {}, total size {}. raw_size_per_file {},duration {:?}",
```
Comment (nolouch) [inline components/raftstore/src/store/snap/io.rs:187]:
```suggestion
            "build_sst_cf_file_list builds {} files in cf {}. Total keys {}, total size {}. raw_size_per_file {}, total tasks {:?}",
```

File: src/server/snap.rs [hunk 2]
```diff
@@ -269,6 +271,7 @@ impl RecvSnapContext {
         if let Err(e) = raft_router.send_raft_msg(self.raft_msg) {
             return Err(box_err!("{} failed to send snapshot to raft: {}", key, e));
         }
+        info!("saving all snapshot files";"snap_key" => %key,"duration" => ?self.start.saturating_elapsed());
```
Comment (nolouch) [inline src/server/snap.rs:274]:
```suggestion
        info!("saving all snapshot files";"snap_key" => %key,"taks" => ?self.start.saturating_elapsed());
```
BTW, Do you run `cargo fmt`?

**Review by nolouch — COMMENTED** [review]
Could you show the metrics?

**Review by lhy1024 — APPROVED** [review]

**Comment by bufferflies** [comment]
PTAL @AndreMouche again

File: components/raftstore/src/store/worker/metrics.rs [hunk 3]
```diff
@@ -10,10 +10,11 @@ use tikv_util::time::Instant;
 make_auto_flush_static_metric! {
     pub label_enum SnapType {
        generate,
-       apply,
+       apply
     }
 
     pub label_enum SnapStatus {
+        total,
```
Comment (nolouch) [inline components/raftstore/src/store/worker/metrics.rs:17]: what's different between `total` and `all`?

Comment (bufferflies) [reply components/raftstore/src/store/worker/metrics.rs:17]: total means that the task is not runing, it needs to wait.
all means that the task is runnig . 
the sub of them is the waiting count .

**Comment by bufferflies** [comment]
PTAL @Connor1996

File: components/raftstore/src/store/worker/metrics.rs (hunk 3 above)
Comment (nolouch) [reply components/raftstore/src/store/worker/metrics.rs:17]: they are not very distinguishable. How about change `all` to `running`? I think origin `all` is unreasonable.

File: components/raftstore/src/store/worker/region.rs [hunk 4]
```diff
@@ -805,7 +805,7 @@ where
                         allow_multi_files_snapshot = !is_tiflash;
                     }
                 }
-
+                SNAP_COUNTER.generate.total.inc();
```
Comment (nolouch) [inline components/raftstore/src/store/worker/region.rs:808]: no  `SNAP_COUNTER.apply.total.inc()`?

Comment (bufferflies) [reply components/raftstore/src/store/worker/region.rs]: fix

File: components/raftstore/src/store/worker/metrics.rs (hunk 3 above)
Comment (bufferflies) [reply components/raftstore/src/store/worker/metrics.rs]: fix

File: src/server/snap.rs [hunk 5]
```diff
@@ -151,7 +151,9 @@ pub fn send_snap(
         return Err(box_err!("missing snap file: {:?}", s.path()));
     }
     let total_size = s.total_size();
-
+    SNAPSHOT_LIMIT_TRANSPORT_BYTES_VEC
+        .with_label_values(&["send"])
```
Comment (Connor1996) [inline src/server/snap.rs:155]: use static metrics

Comment (bufferflies) [reply src/server/snap.rs]: fix

File: components/raftstore/src/store/worker/metrics.rs [hunk 6]
```diff
@@ -10,11 +10,18 @@ use tikv_util::time::Instant;
...
        generate,
-       apply,
+       apply
     }
 
+    // snapshot task status
+    // |all---------start--------------|
+    //               |
+    //               |
+    //               V
```
Comment (Connor1996) [inline components/raftstore/src/store/worker/metrics.rs:20]: what's the meaning of this arrow

File: src/server/metrics.rs [hunk 7]
```diff
@@ -106,6 +106,10 @@ make_auto_flush_static_metric! {
         "type" => SnapTask,
     }
 
+    pub struct SnapLimitTransportBytesVec:LocalIntCounter{
+        "type" =>SnapTask,
```
Comment (Connor1996) [inline src/server/metrics.rs:110]:
```suggestion
        "type" => SnapTask,
```

File: components/raftstore/src/store/metrics.rs [hunk 8]
```diff
@@ -778,4 +778,10 @@ lazy_static! {
         "Sum of applying sst.",
         &["type"]
     ).unwrap();
+
+    pub static ref SNAPSHOT_LIMIT_GENERATE_BYTES: IntCounter = register_int_counter!(
+        "tikv_snapshot_limit_generate_bytes",
```
Comment (Connor1996) [inline components/raftstore/src/store/metrics.rs:783]: where is the limit?

File: components/raftstore/src/store/snap.rs [hunk 9]
```diff
@@ -887,6 +887,7 @@ impl Snapshot {
                     &self.mgr.limiter,
                 )?
             };
+            SNAPSHOT_LIMIT_GENERATE_BYTES.inc_by(cf_stat.total_size as u64);
```
Comment (Connor1996) [inline components/raftstore/src/store/snap.rs:890]: why don't consider plain files?

File: components/raftstore/src/store/worker/metrics.rs (hunk 6 above)
Comment (bufferflies) [reply components/raftstore/src/store/worker/metrics.rs:20]: It means that the trnasfer relation of one snapshot task . It begin at all,  then into  start by executing ,  finally succ. and the abnormal status may be cancel,  abort，fail etc.

File: src/server/metrics.rs (hunk 7 above)
Comment (bufferflies) [reply src/server/metrics.rs:110]: fix

File: components/raftstore/src/store/metrics.rs (hunk 8 above)
Comment (bufferflies) [reply components/raftstore/src/store/metrics.rs:783]: It is the io limit  to limit the snapshot disk. https://github.com/tikv/tikv/blob/da5396b2c2e3ae985cf195f35f52d955d45cd381/components/raftstore/src/store/snap/io.rs#L111-L118

File: components/raftstore/src/store/snap.rs (hunk 9 above)
Comment (bufferflies) [reply components/raftstore/src/store/snap.rs:891]: It only record io limit used.

File: components/raftstore/src/store/worker/region.rs [hunk 10]
```diff
@@ -313,9 +313,10 @@ where
         allow_multi_files_snapshot: bool,
     ) {
         fail_point!("before_region_gen_snap", |_| ());
-        SNAP_COUNTER.generate.all.inc();
+        SNAP_COUNTER.generate.start.inc();
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply components/raftstore/src/store/worker/region.rs:316]: some task maybe need to wait the previous executing finish, like this:
10 tasks need to runnig , every task need 10s , and running 2 task at the same time: so the metrics may be this :
[all: 10 ,start: 2] and later 10s  [all: 10, 4]

File: src/server/snap.rs [hunk 11]
```diff
@@ -470,7 +479,7 @@ where
                             cb(Ok(()));
                         }
                         Err(e) => {
-                            error!("failed to send snap"; "to_addr" => addr, "err" => ?e);
+                            error!("failed to send snap"; "to_addr" => addr, "region_id"=>region_id, "err" => ?e);
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply src/server/snap.rs]: fix

File: src/server/snap.rs [hunk 12]
```diff
@@ -270,6 +274,7 @@ impl RecvSnapContext {
         if let Err(e) = raft_router.send_raft_msg(self.raft_msg) {
             return Err(box_err!("{} failed to send snapshot to raft: {}", key, e));
         }
+        info!("saving all snapshot files";"snap_key" => %key, "takes" => ?self.start.saturating_elapsed());
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply src/server/snap.rs]: fix

**Review by Connor1996 — APPROVED** [review]
LGTM

**Review by nolouch — APPROVED** [review]

**Comment by bufferflies** [comment]
/ping @Connor1996 ， help me merge it

**Review by AndreMouche — APPROVED** [review]

**Pull request merged**